### PR TITLE
Fix the Route of subscription API does not work on pubsub.pulsar

### PR DESCRIPTION
**Pull request by Taction**
# Description

Fix the Route of subscription API does not work on pubsub.pulsar
Add tenant and namespace for pubsub.pulsar

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1263 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#1915

### CODE REVIEW

**Comment by yaron2** [comment]
@daixiang0 can you please review this?

**Comment by daixiang0** [comment]
I would @yaron2 .

File: pubsub/pulsar/pulsar.go [hunk 1]
```diff
@@ -241,3 +256,8 @@ func (p *Pulsar) Close() error {
 func (p *Pulsar) Features() []pubsub.Feature {
 	return nil
 }
+
+// formatTopic the topic into pulsar's format with tenant and namespace.
+func (p *Pulsar) formatTopic(topic string) string {
```
Comment (daixiang0) [inline pubsub/pulsar/pulsar.go:261]: Please add ut for it.

**Comment by beiwei30** [comment]
This change looks good to me too, except for additional doc change, and suggestion from @daixiang0 for more unit tests.

File: pubsub/pulsar/pulsar.go (hunk 1 above)
Comment (Taction) [reply pubsub/pulsar/pulsar.go:261]: done.

File: pubsub/pulsar/pulsar.go [hunk 2]
```diff
@@ -162,32 +176,33 @@ func parsePublishMetadata(req *pubsub.PublishRequest) (
...
+		Topic:            topic,
 		SubscriptionName: p.metadata.ConsumerID,
 		Type:             pulsar.Failover,
 		MessageChannel:   channel,
 	}
 
 	consumer, err := p.client.Subscribe(options)
 	if err != nil {
-		p.logger.Debugf("Could not subscribe %s", req.Topic)
+		p.logger.Debugf("Could not subscribe %s, full topic in pulsar is %s", req.Topic, topic)
```
[the remark this answers is not in the document]
Comment (Taction) [reply pubsub/pulsar/pulsar.go]: done.

File: pubsub/pulsar/pulsar.go [hunk 3]
```diff
@@ -111,17 +124,18 @@ func (p *Pulsar) Publish(req *pubsub.PublishRequest) error {
 		msg      *pulsar.ProducerMessage
 		err      error
 	)
-	cache, _ := p.cache.Get(req.Topic)
+	topic := p.formatTopic(req.Topic)
+	cache, _ := p.cache.Get(topic)
 	if cache == nil {
-		p.logger.Debugf("creating producer for topic %s", req.Topic)
+		p.logger.Debugf("creating producer for topic %s, full topic in pulsar is %s", req.Topic, topic)
```
[the remark this answers is not in the document]
Comment (Taction) [reply pubsub/pulsar/pulsar.go]: done.

File: pubsub/pulsar/pulsar.go [hunk 4]
```diff
@@ -16,11 +16,17 @@ import (
...
+	enableTLS    = "enableTLS"
+	deliverAt    = "deliverAt"
+	deliverAfter = "deliverAfter"
+	tenant       = "tenant"
+	namespace    = "namespace"
+
+	defaultTenant     = "public"
+	defaultNamespace  = "default"
 	cachedNumProducer = 10
+	topicFormat       = "persistent://%s/%s/%s"
```
[the remark this answers is not in the document]
Comment (Taction) [reply pubsub/pulsar/pulsar.go]: Added non-persistent support. Added comment with docs link.

**Comment by mukundansundar** [comment]
overall looks good to me ... one question I have is if non-persistent topic is used, how should the retries be configured in that scenario ? should there be retries for non-persistent topic subscriptions ?
/cc @pkedy @Taction

**Comment by Taction** [comment]
> overall looks good to me ... one question I have is if non-persistent topic is used, how should the retries be configured in that scenario ? should there be retries for non-persistent topic subscriptions ? /cc @pkedy @Taction

@mukundansundar I think the configuration to retries has little to do with whether it is persisted or not. For non-persistent, unconsumed messages will be discarded after disconnection. For retrying, it tries best to allow the service to consume the messages when it is running normally. In addition, changing the default retry behavior when using non-persistent topics may cause confusion to users.

In short, I think it is ok to configure retry for non-persistent messages. After all, retry is configurable, users can disable retry through the retry metadata, I will document that users can disable retry when using non-persistent topics in the documentation.

**Review by mukundansundar — APPROVED** [review]
lgtm

**Review by berndverst — APPROVED** [review]

**Review by artursouza — APPROVED** [review]
